rust-runtime/aws-smithy-http/src/byte_stream.rs

```diff
@@ -326,6 +326,14 @@ impl From<Vec<u8>> for ByteStream {
     }
 }
 
+impl From<hyper::Body> for ByteStream {
```

Comment:
SdkBody actually implements `From<hyper::Body>`. So you can do `ByteStream::new(SdkBody::from(input))`

Reply:
@rcoh that doesn't work because of the boxed error type. I tried that at first.
You can see how I converted from hyper::Body to ByteStream :
```
        ByteStream::new(SdkBody::from_dyn(
            input.map_err(|e| e.into_cause().unwrap()).boxed(),
        ))
```

So I preferred to have that specifically done outside of generated code, but can change to either way.

Reply:
what issues are caused by the box error type?

this seems to compile fine:
```rust
impl From<hyper::Body> for ByteStream {
    fn from(body: hyper::Body) -> Self {
        ByteStream::new(SdkBody::from(body))
    }
}
```

Reply:
Let me capture the error in a few minutes and paste. I think this is related to the generated code trait bounds which are not explicitly using hyper::Body.

Reply:
Here is the generated code for example:
```
#[allow(clippy::unnecessary_wraps)]
pub async fn parse_put_object_request<B>(
    #[allow(unused_variables)] request: &mut axum_core::extract::RequestParts<B>,
) -> std::result::Result<
    crate::input::PutObjectInput,
    aws_smithy_http_server::rejection::SmithyRejection,
>
where
    B: aws_smithy_http_server::HttpBody + Send,
    B: Into<aws_smithy_http::byte_stream::ByteStream>,
    B::Data: Send,
    B::Error: Into<aws_smithy_http_server::BoxError>,
    aws_smithy_http_server::rejection::SmithyRejection:
        From<<B as aws_smithy_http_server::HttpBody>::Error>,
{
    Ok({
        #[allow(unused_mut)]
        let mut input = crate::input::put_object_input::Builder::default();
        let body = request
            .take_body()
            .ok_or(aws_smithy_http_server::rejection::BodyAlreadyExtracted)?;
        input = input.set_body(Some(
            aws_smithy_http::byte_stream::ByteStream::new(
                aws_smithy_http::body::SdkBody::from(body)
            )
        ));
        // input = input.set_body(Some(body.into()));
```

And the error is:
```
error[E0277]: the trait bound `SdkBody: From<B>` is not satisfied
    --> codegen/build/smithyprojections/s3d/s3/rust-server-codegen/src/operation_deser.rs:4173:17
     |
4173 |                 aws_smithy_http::body::SdkBody::from(body)
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<B>` is not implemented for `SdkBody`
     |
help: consider extending the `where` bound, but there might be an alternative better way to express this requirement
     |
4163 |         From<<B as aws_smithy_http_server::HttpBody>::Error>, SdkBody: From<B>
     |                                                             ~~~~~~~~~~~~~~~~~~
```

So instead of the trait bound that I added `B: Into<aws_smithy_http::byte_stream::ByteStream>` I could do `B: Into<aws_smithy_http::body::SdkBody>` and then do as you suggested. The only question is why we need SdkBody here and not skip directly to the stream. CC @david-perez

Reply:
to be honest, the client used to have generic bodies but it ended up being a big hassle. It might be worth just using SdkBody directly and removing the generated generic

Reply:
anyway, maybe I'm missing something but adding the `From<hyper::Body>` doesn't seem related to this function which uses `HttpBody`?

Reply:
I think the confusion is that it doesn't use HttpBody trait, which doesn't have any conversion to a stream, instead it uses the trait bound that I added `B: Into<ByteStream>` which I then had to implement for the hyper::Body struct to make a hyper server compile.

Reply:
In any case I can try any direction you have to improve and simplify, but I think the generic nature of the generated server code that allows any concrete body type that implements the required bounds is the reason for this complexity, so if we want to change the generated server code to fixate on SdkBody struct that will make the code simpler, but less flexible if theres a need to support another server stack.

Reply:
got it! ok I understand now. Here's what I would do:

Being able to construct a ByteStream from HttpBody is generally useful, we should support that. We should add a purpose built convert to create a ByteStream from the server generic—since this will be just be a fat pointer to whatever the server is using, it maintains flexibility with minimal overhead. 

I probably _would not_ change to smithy_http_server::ByteStream but that is up to @david-perez

Reply:
@rcoh 

> I probably would not change to smithy_http_server::ByteStream but that is up to @david-perez

The only reasons why I suggested this is because (1) `ByteStream` is tied to `SdkBody` and (2) we would expose `SdkBody` in the public API and server users will be confused. If we can make `ByteStream` generic over the body type and thus reuse code across client and server that would be perfect.

Reply:
I tried this for the client—it just causes loads of complexity for very little gain (IMO), especially since SDK Body is capable of holding a `dyn Body`.

We _could_ add a generic to ByteStream with a default parameter actually...but I'd need to be a little more convinced of it's utility

Reply:
@rcoh @david-perez 
I don't think I have the context to assist in this decision at this time. 

Would it make sense that for this PR I just keep using ByteStream as is (or re-export as is from smithy-http-server) and we can continue adjusting this in subsequent PRs?

Reply:
What do you say @david-perez ?

Reply:
@david-perez ?

Reply:
@guymguym Sure, we can leave that change for a subsequent PR if it makes it easier for you.

Reply:
That will be good. Thanks

Reply:
@guymguym We have decided that we will use [`aws_smithy_http::byte_stream::ByteStream`](https://docs.rs/aws-smithy-http/latest/aws_smithy_http/byte_stream/struct.ByteStream.html), making all methods referencing `SdkBody` `#[doc(hidden)]` (`new`, `into_inner`, `From<SdkBody>` impl), and we will add generic methods that don’t reference `SdkBody`.

cc @crisidev @rcoh

Reply:
@david-perez Sounds good. 
Should I change anything in this `impl From<hyper::Body> for ByteStream` code?